### PR TITLE
NT-1763:Refactor CommentsActivity

### PR DESCRIPTION
**Pull request by hadia**
# 📲 What
Remove butter knife from CommentsActivity and replaced with view binding

# 🤔 Why

deprecated butterknife

# 🛠 How

1- change CommentsActivity from java to kotlin
3-remove butter-knife and replace it with viewbinding

# 👀 See

https://kickstarter.atlassian.net/browse/NT-1763

| Before 🐛 | After 🦋 |

https://user-images.githubusercontent.com/1075310/107567720-b64fbe80-6bee-11eb-9724-2d65188f1c16.mp4

# Story 📖

https://kickstarter.atlassian.net/browse/NT-1763

### CODE REVIEW

File: app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        if (postCommentButton != null) {
+            postCommentButton.isEnabled = enabled
+        }
+    }
+
+    override fun projectContextClicked(viewHolder: ProjectContextViewHolder?) {
+        back()
+    }
+
+    override fun emptyCommentsLoginClicked(viewHolder: EmptyCommentsViewHolder?) {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt:173]: nit picky detail, inline function here `override fun emptyCommentsLoginClicked(viewHolder: EmptyCommentsViewHolder?) = commentsLogin()`

File: app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        alertDialog.onNext(commentDialog)
+    }
+
+    fun setPostButtonEnabled(postCommentButton: TextView?, enabled: Boolean) {
+        if (postCommentButton != null) {
+            postCommentButton.isEnabled = enabled
+        }
+    }
+
+    override fun projectContextClicked(viewHolder: ProjectContextViewHolder?) {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt:169]: nit picky detail, inline function

File: app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        /* Toolbar UI actions */
+        val projectNameTextView = commentDialog.findViewById<TextView>(R.id.comment_project_name)
+        projectNameTextView?.text = project.name()
+
+        // Handle cancel-click region outside of dialog modal.
+        commentDialog.setOnCancelListener { dialogInterface: DialogInterface -> viewModel.inputs.commentDialogDismissed() }
+        alertDialog.onNext(commentDialog)
+    }
+
+    fun setPostButtonEnabled(postCommentButton: TextView?, enabled: Boolean) {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt:163]: nit picky detail, inline function`fun setPostButtonEnabled(postCommentButton: TextView?, enabled: Boolean) = postCommentButton?.let { it.isEnabled = enabled }`

File: app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        back()
+    }
+
+    private fun commentsLogin() {
+        val intent = Intent(this, LoginToutActivity::class.java)
+            .putExtra(IntentKey.LOGIN_REASON, LoginReason.COMMENT_FEED)
+        startActivityForResult(intent, ActivityRequestCodes.LOGIN_FLOW)
+    }
+
+    private fun commentButtonClicked() {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt:139]: nit picky detail, same for the inline functions for this method and for `dismissCommentDialog` and `projectContextViewClick`

File: app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+import com.kickstarter.ui.viewholders.ProjectContextViewHolder
+import com.kickstarter.viewmodels.CommentsViewModel
+import com.trello.rxlifecycle.ActivityEvent
+import rx.Observable
+import rx.android.schedulers.AndroidSchedulers
+import rx.subjects.PublishSubject
+
+@RequiresActivityViewModel(CommentsViewModel.ViewModel::class)
+class CommentsActivity : BaseActivity<CommentsViewModel.ViewModel>(), CommentsAdapter.Delegate {
+    private var adapter: CommentsAdapter? = null
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt:38]: these 3 optionals can be changed to latinits
`  private lateinit var adapter: CommentsAdapter
    private lateinit var recyclerViewPaginator: RecyclerViewPaginator
    private lateinit var swipeRefresher: SwipeRefresher`

File: app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            .map { it?.findViewById<TextView>(R.id.comment_body) }
+        val postCommentButton = alertDialog
+            .map { it?.findViewById<TextView>(R.id.post_button) }
+        val cancelButton = alertDialog
+            .map { it?.findViewById<TextView>(R.id.cancel_button) }
+        cancelButton
+            .switchMap { view -> view?.let { RxView.clicks(it) } }
+            .observeOn(AndroidSchedulers.mainThread())
+            .compose(bindToLifecycle())
+            .subscribe { viewModel.inputs.commentDialogDismissed() }
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt:66]: nit picky detail, can we let a line separation between each subscription? it's hard to read without separations

File: app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+            this, binding.commentsSwipeRefreshLayout, { viewModel.inputs.refresh() }
+        ) { viewModel.outputs.isFetchingComments }
+        val commentBodyEditText = alertDialog
+            .map { it?.findViewById<TextView>(R.id.comment_body) }
+        val postCommentButton = alertDialog
+            .map { it?.findViewById<TextView>(R.id.post_button) }
+        val cancelButton = alertDialog
+            .map { it?.findViewById<TextView>(R.id.cancel_button) }
+        cancelButton
+            .switchMap { view -> view?.let { RxView.clicks(it) } }
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt:63]: Name shadowed `view`

File: app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            .observeOn(AndroidSchedulers.mainThread())
+            .compose(bindToLifecycle())
+            .subscribe { viewModel.inputs.commentDialogDismissed() }
+        postCommentButton
+            .switchMap { view -> view?.let { RxView.clicks(it) } }
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt:68]: Name shadowed `view`

**Review by Arkariang — CHANGES_REQUESTED** [review]
Great work here 👍 ! 
From my comments the only ones I'm requesting changes is for the optionals to be lateinit instead, and the shadowed names. The rest of them is a matter of preference on the syntax.

Thanks for the small PR! is way easier to review :)

**Review by Arkariang — APPROVED** [review]
Great work @hadia !

File: app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt (hunk 6 above)
Comment (leighdouglas) [reply app/src/main/java/com/kickstarter/ui/activities/CommentsActivity.kt:66]: +1

**Comment by leighdouglas** [comment]
Great work! I agree with @Arkariang 's comments about style and syntax. Once these changes are made I can approve 😀

**Review by leighdouglas — APPROVED** [review]
LGTM! 🎉

**Pull request merged**